R/visualize_embeddings.R

```diff
@@ -143,77 +144,150 @@ keep_tokens <- function(embedding_df, tokens = "[CLS]") {
+  )
 }
 
+# do_pca -------------------------------------------------------------
 
+#' Do PCA on Embeddings
+#'
+#' Calculate the 2D PCA for a collection of embedding vectors, in preparation
+#' for plotting.
+#'
+#' @inheritParams display_pca
+#'
+#' @return A tbl_df of the embedding vectors projected onto two principal
+#'   axes.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # assuming something like the following has been run:
+#' # feats <- RBERT::extract_features(...) # See RBERT documentation
+#' # Then:
+#' pca_df <- feats$output %>%
+#'     filter_layer_embeddings(layer_indices = 12L) %>%
+#'     keep_tokens("[CLS]") %>%
+#'     do_pca()
+#' }
+do_pca <- function(embedding_df,
+                   project_vectors = embedding_df,
+                   color_field = NULL,
+                   disambiguate_tokens = TRUE) {
+  num_rows <- nrow(embedding_df)
+  num_rows_proj <- nrow(project_vectors)
+  if (num_rows_proj < 3) {
+    stop("At least three vectors are required for a ",
+         "meaningful PCA plot.")
+  }
+  # Use just the indicated vectors to select the PCA projection.
+  proj_mat <- as.matrix(
+    dplyr::select(project_vectors,
+                  dplyr::matches("V[0-9]+"))
+  )
+  pcs <- stats::prcomp(proj_mat,
+                       retx = TRUE, center = TRUE, scale. = TRUE, rank. = 2L)
+
+  if (disambiguate_tokens) {
+    embedding_df <- dplyr::mutate(embedding_df,
+                                  token = paste(token,
+                                                sequence_index,
+                                                token_index,
+                                                sep = "."))
+  }
+  # Keep all the non-vector columns, for possible use in plotting.
+  tok_labels <- dplyr::select(embedding_df,
+                              -dplyr::matches("V[0-9]+"))
+
+  vec_mat <- as.matrix(
+    dplyr::select(embedding_df,
+                  dplyr::matches("V[0-9]+"))
+  )
+  # Instead of doing PCA here, apply the scaling and projection manually.
+  vec_mat <- scale(vec_mat, center = pcs$center, scale = pcs$scale)
+  projected <- vec_mat %*% pcs$rotation
+
+  pca_df <- dplyr::bind_cols(tok_labels, tibble::as_tibble(projected))
+  return(pca_df)
+}
+
+# plot_pca -------------------------------------------------------------
+
+#' Plot PCA
+#'
+#' Given the output of \code{prep_pca}, make the plot object.
+#'
+#' @param pca_df A tbl_df of pca vectors; output from a call to \code{prep_pca}.
+#' @inheritParams display_pca
+#'
+#' @return A ggplot2 plot of the embedding vectors projected onto two principal
+#'   axes.
+#' @export
+#'
+#' @examples
+#' \dontrun{
+#' # assuming something like the following has been run:
+#' # feats <- RBERT::extract_features(...) # See RBERT documentation
+#' # Then:
+#' feats$output %>%
+#'     filter_layer_embeddings(layer_indices = 12L) %>%
+#'     keep_tokens("[CLS]") %>%
+#'     prep_pca() %>%
+#'     plot_pca()
+#' }
+plot_pca <- function(pca_df,
```

Comment:
What if... we rename display_pca to plot_pca, add a class when we calculate_pca, and then make this a method of the main function (with the current main function becoming the default)? Possible big negative: the main function would have ... params (cuz the params change depending on the class), so.... maybe not?

Let's discuss in Slack, I'm not quite nailing this...

Reply:
I don't think there's a need to mess around with classes for this. Maybe down the road, if we have other candidates for classes? For now, I don't think it's worth the overhead.

Reply:
Yeah, that's fine. I think we just need to resolve the naming issue.